providers/vic/pod_deleter.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2018 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vic
+
+import (
+	"github.com/virtual-kubelet/virtual-kubelet/providers/vic/cache"
+	vicpod "github.com/virtual-kubelet/virtual-kubelet/providers/vic/pod"
+	"github.com/virtual-kubelet/virtual-kubelet/providers/vic/proxy"
+	"github.com/vmware/vic/lib/apiservers/portlayer/client"
+	"github.com/vmware/vic/pkg/trace"
+
+	"k8s.io/api/core/v1"
+	"fmt"
+)
+
+type PodDeleter interface {
+	DeletePod(op trace.Operation, pod *v1.Pod, start bool) error
+}
+
+type VicPodDeleter struct {
+	client         *client.PortLayer
+	imageStore     proxy.ImageStore
+	isolationProxy proxy.IsolationProxy
+	podCache       cache.PodCache
+	personaAddr    string
+	portlayerAddr  string
+}
+
+type DeleteResponse struct {
+	Id       string `json:"Id"`
+	Warnings string `json:"Warnings"`
+}
+
+func NewPodDeleter(client *client.PortLayer, isolationProxy proxy.IsolationProxy, podCache cache.PodCache, personaAddr string, portlayerAddr string) *VicPodDeleter {
+	return &VicPodDeleter{
+		client:         client,
+		podCache:       podCache,
+		personaAddr:    personaAddr,
+		portlayerAddr:  portlayerAddr,
+		isolationProxy: isolationProxy,
+	}
+}
+
+func (v *VicPodDeleter) DeletePod(op trace.Operation, pod *v1.Pod) error {
+	defer trace.End(trace.Begin(pod.Name, op))
+
+	// Get pod from cache
+	vp, err := v.podCache.Get(op, "", pod.Name)
+
+	if err != nil {
+		return err
+	}
+
+	// Stop pod if not already stopped
+
```

Comment:
I think we need to have a lock here to protect the vm deletion and the cache deletion.